File: Chapitre3/exercice3.5.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/wait.h>

int main() {
    char nom_fichier[256];

    while (1) {
        // 1. Saisie du nom de fichier
        printf("\nEntrez le nom du fichier à ouvrir (ou 'q' pour quitter) : ");
        scanf("%s", nom_fichier);

        // 2. Condition de sortie
        if (strcmp(nom_fichier, "q") == 0) {
            break;
        }

        // 3. Création d'un processus fils
        // Indispensable car exec remplace le programme en cours
        pid_t pid = fork();

        if (pid == -1) {
            perror("Erreur fork");
            exit(1);
        }

        if (pid == 0) {
            // Dans le processus fils : on se transforme en emacs
            // On utilise le chemin complet vers l'exécutable [5]
            execl("/usr/bin/vim", "vim", nom_fichier, NULL);

            // Si execl échoue (ex: emacs non installé à cet endroit)
            perror("Erreur lors de l'ouverture d'emacs");
            exit(1);
        } else {
            // Dans le processus père : on attend la fermeture d'emacs
            // pour pouvoir saisir le nom du fichier suivant [6]
            wait(NULL);
            printf("Fermeture d'emacs détectée. Prêt pour un nouveau fichier.\n");
        }
    }

    printf("Fin du programme.\n");
    return 0;
}

File: Chapitre3/exercice3.5.py
import os
import sys

def main():

    while True:
        # ① Saisir le nom du fichier
        nom_fichier = input("\nEntrez le nom du fichier à ouvrir "
                           "(ou 'q' pour quitter) : ").strip()

        # ② Condition de sortie
        if nom_fichier == "q":
            break

        # ③ Vérifier que le nom n'est pas vide
        if not nom_fichier:
            print("Nom vide, veuillez recommencer.")
            continue

        # ④ Créer un fils pour lancer vim
        pid = os.fork()

        if pid == -1:
            print("Erreur fork")
            sys.exit(1)

        if pid == 0:
            # ── FILS : lancer vim
            os.execl("/usr/bin/vim", "vim", nom_fichier)

            # Atteint seulement si execl échoue
            print("Erreur lors de l'ouverture de vim")
            sys.exit(1)

        else:
            # ── PÈRE : attendre la fermeture de vim
            pid_termine, status = os.waitpid(pid, 0)

            if os.WIFEXITED(status) and os.WEXITSTATUS(status) == 0:
                print("Fermeture de vim détectée. "
                      "Prêt pour un nouveau fichier.")
            else:
                print("vim s'est terminé avec une erreur.")

    print("\nFin du programme.")

if __name__ == "__main__":
    main()